src/Extensions/Microsoft.Diagnostics.Monitoring.Extension.Common/SharedEntrypoint.cs

```diff
@@ -0,0 +1,122 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+using System;
+using System.Collections.Generic;
+using System.CommandLine;
+using System.CommandLine.Parsing;
+using System.IO;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Monitoring.Extension.Common
+{
+    internal sealed class SharedEntrypoint<T> where T : class, IEgressProviderOptions, new()
+    {
+        private static Stream StdInStream;
+        private static CancellationTokenSource CancelSource = new CancellationTokenSource();
+
+        private static EgressProvider _provider;
+        private static Action<ExtensionEgressPayload, T> _configureOptions;
+
+        public static async Task<int> Entrypoint(string[] args, EgressProvider provider, Action<ExtensionEgressPayload, T> configureOptions)
+        {
+            _provider = provider;
+            _configureOptions = configureOptions;
+
+            using var loggerFactory = LoggerFactory.Create(builder =>
+            {
+                builder.AddConsole();
+            });
+
+            // Expected command line format is: dotnet-monitor-egress-*extension_type*.exe Egress
+            RootCommand rootCommand = new RootCommand("Egresses an artifact to cloud storage.");
+
+            Command egressCmd = new Command("Egress", "The class of extension being invoked; Egress is for egressing an artifact.");
+
+            egressCmd.SetHandler(Egress);
+
+            rootCommand.Add(egressCmd);
+
+            return await rootCommand.InvokeAsync(args);
+        }
+
+        private static async Task<int> Egress()
+        {
+            EgressArtifactResult result = new();
+            try
+            {
+                string jsonConfig = Console.ReadLine();
+                ExtensionEgressPayload configPayload = JsonSerializer.Deserialize<ExtensionEgressPayload>(jsonConfig);
+                IEgressProviderOptions options = BuildOptions(configPayload);
+
+                Console.CancelKeyPress += Console_CancelKeyPress;
```

Comment:
I'd like to see the cancellation hook moved to a higher layer, as well as the source of the stdin and stdout streams; this will allow for easier testing of components. This does not have to be done in this change.